Add some documentation (#110)
square/Aardvark: Documentation/Logging.md

# Logging

One of the most helpful components of an actionable bug report is a complete set of logs. Aardvark makes it simple to add logs with the logging utilities in the CoreAardvark framework.

## Getting Started

The easiest way to get started logging with Aardvark is by replacing calls to `print` with `log` in Swift and calls to `NSLog` with `ARKLog` in Objective-C.

```swift
log("Hello world")
```

```objc
ARKLog(@"Hello world");
```

By default, this will log messages to the Aardvark log store _instead of_ the console. If you would like the messages to be logged to the console as well, toggle the default log store's `printsLogsToConsole` property:

```swift
ARKLogDistributor.default().defaultLogStore.printsLogsToConsole = true
```

## Adding Parameters

> A recommended pattern for logging is to keep your log messages simple and attach details of the log in parameters. You can attach key/value pairs of strings to your log messages to provide this extra detail, while keeping your logs easy to scan through in a list of messages.

Reviewer: It's not clear what "parameters" is as-is on the first read of this first sentence. I tried to make it slightly more clear. I also tried to clarify the point of "extra detail". Feel free to take some or none of this suggestion:
```suggestion
A recommended pattern for logging is to keep your log messages simple and attach details of the log in the `parameters` field. You can attach key/value pairs of strings to your log messages to provide further debugging data, while keeping your logs easy to scan through in a list of messages.
```

Author: I like it 👍

```swift

> log("Said hello to user", parameters: ["user_name": user.name])

Reviewer: This example rocks!

```

```objc
ARKLogWithParameters(@{ @"user_name": [user name] }, @"Said hello to user");
```

## Using Dependency Injection

> If your app is architected around dependency injection, you can inject an `ARKLogDistributor` instead of using the global logging functions.

Reviewer: Maybe:
```suggestion
If prefer to use dependency injection rather than global functions, you can inject an `ARKLogDistributor` to your logging call-sites.
```

Context: Airbnb's app has dependency injection, but the one thing we use a (runtime-set) singleton for is logging. Sometimes mixed worlds exist 🙂

Author: Ahh good call out. Updated.

```swift
func sayHello(to name: String, on logDistributor: ARKLogDistributor) {
    logDistributor.log("Hello \(name)")
}
```

```objc

> - (void)sayHelloToPersonNamed:(NSString *)name onLogDistributor:(ARKLogDistributor *)logDistributor;

Reviewer: extreme nit: the signature of this method is not the same as the signature of the swift method above. They don't have to be. But...
![giphy](https://user-images.githubusercontent.com/139364/108608628-1aa41800-737d-11eb-8ab2-fe5546ce78c5.gif)

Author: 😂 Fair enough. Updated the Swift signature.

{
    [logDistributor logWithFormat:@"Hello %@", name];
}
```

## Using Multiple Log Stores

> By default, all logs do into the same log file. If you would prefer to separate logs into different files, you can add multiple log stores to the distributor and set each store's `logFilterBlock` to filter out which logs are included.

Reviewer: In the first sentence, I think you meant "go" instead of "do"?

In any case, I reworded slightly:
```suggestion
By default, all logs are sent to the same log file. If you would prefer to send some logs into separate log files, you can add multiple log stores to the distributor and set each store's `logFilterBlock` to filter incoming logs.
```

> To facilitate this, log messages contain a `userInfo` dictionary that can be used to control behavior in log observers. Unlike the `parameters` dictionary, the `userInfo` is not persisted with the log message.

Reviewer: let's avoid using "this" without a very clear antecedent. I also deleted language that re-described the problem described above.
```suggestion
To facilitate filtering, log messages contain a `userInfo` dictionary. Unlike the `parameters` dictionary, the `userInfo` is not persisted with the log message.
```

```swift
log("Loaded data", userInfo: ["category": "Core Data"])
```

> For example, we can set up a second log store for all of the logs related to Core Data that is separate from the rest of the logs. This could be helpful if our Core Data implementation consistently runs tasks in the background and creates noise in the main log store.

Reviewer: we've already said the log store is a second one, so instead of repeating that the logs are separate let's be explicit that we're only sending those logs to one store?
```suggestion
For example, we can set up a second log store for all of the logs related to Core Data and filter these logs out of the main log store. This could be helpful if our Core Data implementation consistently runs tasks in the background and creates noise in the main log store.
```

```swift
// Create the new log store.
let coreDataLogStore = ARKLogStore(persistedLogFileName: "CoreDataLogs.data")

// Set the new log store to only include logs related to Core Data.
coreDataLogStore.logFilterBlock = { message in

>     return message.userInfo?["category"] == "Core Data"

Reviewer: no need for the leading return
```suggestion
    message.userInfo?["category"] == "Core Data"
```

Author: Since we still support Xcode 11, I think it's best avoid implicit returns so consumers can copy the example code directly.

}

// Add the new log store to the default distributor.
ARKLogDistributor.default().add(coreDataLogStore)

// Set the default log store to exclude log related to Core Data.
ARKLogDistributor.default().defaultLogStore.logFilterBlock = { message in

>     return message.userInfo?["category"] != "Core Data"

Reviewer: We can avoid repeated logic:
```suggestion
    return !coreDataLogStore.logFilterBlock(message)
```

Reviewer: also let's remove the leading `return` 😉 
```suggestion
    !coreDataLogStore.logFilterBlock(message)
```
